application/F3DStarter.cxx

```diff
@@ -695,6 +703,141 @@ void F3DStarter::Render()
+    }
+    {
+      std::regex re("\\{version\\}");
+      filename = std::regex_replace(filename, re, F3D::AppVersionFull);
+    }
+    {
+      std::regex re("\\{os\\}");
+      filename = std::regex_replace(filename, re, getPlatform());
+    }
+    {
+      std::regex re("\\{model\\}");
+      if (std::regex_search(filename, re))
+      {
+        const auto file = this->Internals->FilesList[this->Internals->CurrentFileIndex];
+        filename = std::regex_replace(filename, re, file.filename().string());
+      }
+    }
+
+    return std::filesystem::path(filename);
+  };
+
+  const auto findAvailableFilename = [&](const std::string& idealFilename)
+  {
+    const auto appendFilenameNumber = [](const std::filesystem::path& path, size_t n)
+    {
+      std::stringstream ss;
+      const std::string fn = path.string();
+      const std::string::size_type i = fn.find_last_of(".");
+      if (i != std::string::npos)
+      {
+        ss << fn.substr(0, i) << " (" << n << ")" << fn.substr(i);
+      }
+      else
+      {
+        ss << fn << " (" << n << ")";
+      }
+      return ss.str();
+    };
+
+    std::string fn = idealFilename;
+    for (size_t i = 2; std::filesystem::exists(fn) && i < 1e8; ++i)
+    {
+      fn = appendFilenameNumber(idealFilename, i);
+    }
+    if (std::filesystem::exists(fn))
+    {
+      throw std::runtime_error("could not find available filenamne");
+    }
+    return fn;
+  };
+
+  std::filesystem::path path = applyFilenameTemplate(filenameTemplate);
+  if (!path.is_absolute())
+  {
+    path = getScreenshotDir() / path;
+  }
+  path = findAvailableFilename(path);
+  std::filesystem::create_directories(std::filesystem::path(path).parent_path());
+  f3d::log::info("saving screenshot to " + path.string());
+
+  f3d::window& window = this->Internals->Engine->getWindow();
+
+  std::stringstream cameraMetadata;
+  {
+    const auto state = window.getCamera().getState();
+    const auto vec3toJson = [](const std::array<double, 3>& v)
+    {
+      std::stringstream ss;
+      ss << "[" << v[0] << ", " << v[1] << ", " << v[2] << "]";
+      return ss.str();
+    };
+    cameraMetadata << "{\n";
+    cameraMetadata << "  \"pos\": " << vec3toJson(state.pos) << ",\n";
+    cameraMetadata << "  \"foc\": " << vec3toJson(state.foc) << ",\n";
+    cameraMetadata << "  \"up\": " << vec3toJson(state.up) << ",\n";
+    cameraMetadata << "  \"angle\": " << state.angle << "\n";
+    cameraMetadata << "}\n";
+  }
+
+  window.renderToImage()
+    .setMetadata("camera", cameraMetadata.str())
+    .save(path, f3d::image::SaveFormat::PNG);
+
+  f3d::options& options = this->Internals->Engine->getOptions();
+  const std::string light_intensity_key = "render.light.intensity";
+  const double intensity = options.getAsDouble(light_intensity_key);
+  options.set(light_intensity_key, intensity * 5);
+  this->Render();
+  options.set(light_intensity_key, intensity);
+  this->Render();
```

Comment:
This flashes the model white for 1 frame to provide visual feedback.
Not sure if it's the right thing to do? if it is we can do it with a properly timed loop instead, otherwise we can just ditch it